### PR TITLE
Enable code coverage

### PR DESCRIPTION
**Pull request by QuLogic**
Let's see if this works without me setting up anything; then we'll see if it's useful...

### CODE REVIEW

File: ci/make_coverage_siteconfig.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+import os
+import site
+
+try:
+    os.makedirs(site.USER_SITE)
+except OSError:
+    pass
+
+with open(os.path.join(site.USER_SITE, 'sitecustomize.py'), 'w') as f;
```
Comment (ignatenkobrain) [inline ci/make_coverage_siteconfig.py]: __[flake8]__
* [E703] statement ends with a semicolon
* [E999] SyntaxError: invalid syntax

File: ci/make_coverage_siteconfig.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+    f.write('import coverage; coverage.process_startup()')
```
Comment (ignatenkobrain) [inline ci/make_coverage_siteconfig.py]: __[flake8]__
* [E113] unexpected indentation

**Comment by MathieuDuponchelle** [comment]
Might want to document how to run coverage tests locally ?

**Comment by QuLogic** [comment]
Aha, finally. Docker is a bit of a pain sometimes... Now to add Windows.

**Comment by jpakkane** [comment]
This message is posted to every outstanding merge request.

We have transitioned from Github wiki to our new documentation system that generates all documentation directly from the Git repo. This means that from now on every merge request must come with corresponding documentation changes.

If your MR requires documentation changes, do the necessary changes to files in `docs/markdown`.

If your MR adds a new feature, add a description to the release notes in `docs/markdown/Release-notes-for-0.41.0.md`.

If this commit is just a bugfix with no functional changes, you don't need to do anything.

Thank you

**Comment by QuLogic** [comment]
OK, just cygwin left. @jon-turney Do I need to run all my Python stuff through Cygwin's bash? Does Cygwin Python have pip installed? I tried to install it manually, but now I'm not sure if I was just hitting the system Python instead of Cygwin's.

**Comment by jon-turney** [comment]
> OK, just cygwin left. @jon-turney Do I need to run all my Python stuff through Cygwin's bash? Does Cygwin Python have pip installed? I tried to install it manually, but now I'm not sure if I was just hitting the system Python instead of Cygwin's and that's why nothing was working.

python3 is a cygwin symlink to something like python3.6m.exe, so using bash is probably the easiest way to navigate that correctly.

Adding python3-pip to the list of packages to install in ci/appveyor-install.bat should get you pip

**Comment by QuLogic** [comment]
OK, this is 95% there. The outstanding issue is that the top-level scripts are not counted on codecov, though they do appear locally. I'd have to go through a bit more debugging to figure out what that problem is and I think I can leave that for a separate PR.

The last question is how to configure this. The default is to comment on the PR once available, and when any new commits are made, the original comment is edited. We can leave this on, make it "spammy", or turn it off completely. The other configuration is target coverage level; default is not-less-than-previous, but we can set an absolute value

**Comment by nirbheek** [comment]
> The default is to comment on the PR once available, and when any new commits are made, the original comment is edited. We can leave this on, make it "spammy", or turn it off completely.

The default makes sense here.

>The other configuration is target coverage level; default is not-less-than-previous, but we can set an absolute value

I think the default makes sense here again. We will sometimes have PRs that will inevitably reduce coverage because that code is difficult to test for (or is run on platforms that we don't have CI for, etc), but our default should be "any new code must improve the coverage or keep it the same".

**Comment by QuLogic** [comment]
Rebased to fix conflicts and added the badge on there.

**Comment by QuLogic** [comment]
Rebased to fix conflicts.

File: .travis.yml [hunk 3]
```diff
@@ -32,13 +32,27 @@ before_install:
   - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then brew update; fi
   - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then brew install ninja python3; fi
   - if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then docker pull jpakkane/mesonci:zesty; fi
+  - pip3 install codecov
```
Comment (jpakkane) [inline .travis.yml]: Let's not install stuff with pip on every install. I'll update the CI image to contain codecov instead.

Same for other places where pip is used.

Comment (QuLogic) [inline .travis.yml]: Yes, I figured you would do so once this was complete. But we still need this one for osx and uploading.

**Comment by jpakkane** [comment]
The CI image now has codecov.

**Comment by QuLogic** [comment]
Removed the extra install.

**Comment by jpakkane** [comment]
Is this now ready for merging?

**Comment by QuLogic** [comment]
Should be good to go.

File: .travis.yml (hunk 3 above)
Comment (jpakkane) [inline .travis.yml]: It should be run only on OSX then. The Linux container should have everything by default.

Comment (QuLogic) [inline .travis.yml]: Uploading is done in `after_success`, which is outside the container AFAIK.

Comment (jpakkane) [inline .travis.yml]: Ah, right.

**Pull request merged**